### ISSUE TITLE
Testing STM for atomicity fails

### ISSUE THREAD

**Issue by alextoth** [zio/zio#4786]
I designed a test to verify that the atomicity of transactions is preserved inside the STM. This is a followup on #4648.
If atomicity is not preserved, It can not be compensated later on and lead to a false negative. Said differently succeeding failures can not cancel each other. Such situations would lead to think the test passed when it did not.

The reason why this test fails has to do with `zio.internal.Sync` (a wrapper to native synchronised) and the interaction with either the compiler or the JVM.

As evidence of something is wrong (aside of this test):
A/ you can in the code of ZSTM replace `globalLock` value `new AnyRef` with `null` this should trigger a NPE each time the lock is acquired => It never triggers a NPE.   
```scala
val globalLock: AnyRef = null
```
B/ When placing breakpoints with the debugger inside the `Sync(globalLock){` sections, you will see after a few iterations that multiple threads are at the breakpoints inside the critical sections, which is impossible if the sections where locked by `globalLock`.

To solve this the only thing that worked for me  is : 
A/ replace `Sync(globalLock)` by its less magical  counterpart `globalLock.synchronized` 

I guess this is not an issue inside ZIO but inside the compile/exec chain. 

My environment is: scala-2.13.5 && ( java8 || java11.0.2 || java15.0.2) on intel platform macOS

I hope I'm doing something wrong.

```scala
object STM_issue_globalLock_tref extends zio.App {

  import zio.stm.{TRef, ZSTM}
  import zio.{ZIO, _}

  val max = 99

  def run(args: List[String]) = {
    /**
     * If the transaction is executed atomically  it is :
     * sum = sum + ref + i
     * ref = i
     */
    def transaction(i: Int, ref: TRef[Int], sum: TRef[Int]) =
      for {
        ref_prev <- ref.get
        sum_prev <- sum.get
        _ <- sum.set(sum_prev + ref_prev + i)
        _ <- ref.set(i)
      } yield i

    (for {
      (sum, ref) <- ZSTM.atomically(TRef.make[Int](0) zip TRef.make[Int](0))

      fibers <- ZIO.forkAll {
        scala.util.Random.shuffle(1 to max).map { i => ZSTM.atomically(transaction(i, ref, sum)).fork }
      }
      _ <- fibers.join
      (sum, ref) <- ZSTM.atomically {
        sum.get zip ref.get
      }
      passed = sum + ref == (max + 1) * max
      _ <- console.putStrLn(s"passed: $passed :${sum + ref} sum:$sum ref:$ref")
    } yield 0).exitCode
  }
}
```

As a SPEC:
```scala 
        suite("issue #4786")(
          testM("check atomicity ") {
            val max = 99
            def transaction(i: Int, ref: TRef[Int], sum: TRef[Int]) =
              for {
                ref_prev <- ref.get
                sum_prev <- sum.get
                _ <- sum.set(sum_prev + ref_prev + i)
                _ <- ref.set(i)
              } yield i

            for {
              (sum, ref) <- ZSTM.atomically(TRef.make[Int](0) zip TRef.make[Int](0))

              fibers <- ZIO.forkAll {
                scala.util.Random.shuffle(1 to max).map { i => ZSTM.atomically(transaction(i, ref, sum)).fork
                }
              }
              _ <- fibers.join
              (sum, ref) <- ZSTM.atomically {
                sum.get zip ref.get
              }
              passed = sum + ref == (max + 1) * max
            } yield assert(passed)(isTrue)
          } @@ nonFlaky
        )
```

TL;DR

This is a followup on #4648

The way the test works is : [1+2+3+4+...](https://en.wikipedia.org/wiki/1_%2B_2_%2B_3_%2B_4_%2B_⋯) known as [Triangular Numbers](https://en.wikipedia.org/wiki/Triangular_number)
we count the series twice once with `i` and a second time with `ref`.
`ref` is shifted by one iteration so first addition is done with `ref=0` we then need to take the last iteration of `ref` to complete the `sum = (max+1)*max`.

This adds up to `sum+ref = (max+1)*max` at the end of all iterations `1 to max`.
If in the middle of (A)`ref.get  ... ref.set(i)` a concurrent fiber/thread runs (B)`ref.get ... ref.set(i)` 
then at the next iteration the (B) value of `ref` will be lost as it is overwritten by (A) value of `ref`. As a consequence the sum can not match the expected result `(max+1)*max-ref` anymore.
It is impossible to recover accidentally as each transaction has a different `i` and this is guaranteed.

Random.shuffle serves  no purpose to the test only to the reader who may think it works only because numbers are in ascending order.

**Comment by mijicd** [comment]
I'll take a look today!

**Comment by alextoth** [comment]
Thanks a lot @mijicd and @adamgfraser for the details and your time.
I'm sorry this is a clear mistake on my part.

c.f. to #4788 why this was wrong (nested `forkAll/fork` instead of `forkAll` or `forkall/forkDaemon` )
